### ISSUE TITLE
Change workflow from rebase-based to no-rebase merge-based

### ISSUE THREAD

**Issue by ZyX-I** [neovim/neovim#6435]
History: in #6431 it suddenly appeared that almost no commits from #5119 compile because rebases turned previously more or less sane commits into something that has nothing to do with reality: most of these commits were checked neither by CI nor by me and were created automatically by git which does not know what is going on semantically in those commits. This resulted in `git bisect` being actually useless.

Pros:

1. Commits are not loosing context, you get exactly the commit author had when creating it.
2. Semantic rebase/merge failures (e.g. when some merged to master PR changed function name and unmerged rebased PR *added* a line with this name git may not find any conflicts, yet new commit is incorrect) will be limited to just merge commit, no commits if PR author checks merge result locally before recording a merge.
3. Merge conflicts need to be resolved only once. When rebasing it may easily appear that to solve merge conflict you need to edit code which will be altered in some of the next commits resulting in *another* merge conflict.

Cons:

1. History is less nice and would have a bunch of merges.
2. `rebase --interactive` will require a special alias.
3. This is really not a workflow expected by a git users: see the below very lengthy alias, this thing is doing basically the same thing `hg histedit` is: edit history without rebasing. But this functionality is even nearly not in core.
4. If check for “no rebase happened” is implemented fixing CI failure will not be obvious. Though I think I can create a script which will create a script which will fix the problem in many cases after being run by the contributor (but not when contributor actively squashed commits and edited commit messages during rebase), or, at least, restore previous branch state, leaving incorrect branch named `old-name-INCORRECT` for cherry-picking.

The suggestion:

1. PRs should not be rebased now.
2. When conflicting changes appear in master they are merged in.
3. It is allowed to edit history after the latest merge or at the branch point.
4. To edit history after the latest merge/branch point, but *not* rebase (moving the branch point in process) it is suggested to use the following git alias:

        ri = "!sh -c 't=\"${1:-master}\"; s=\"${2:-HEAD}\" ; [ $# -gt 0 ] && shift ; [ $# -gt 0 ] && shift ; git rebase --interactive \"$(git merge-base --is-ancestor \"$t\" \"$s\" && git rev-parse \"$t\" || git merge-base \"$t\" \"$s\")\" \"$@\"' -"

    when PR branch is an active branch this alias may be used as `git ri` which is `rebase --interactive master`, but without rebasing.
5. It is possible to check whether no rebase happened: for all commits `master..pr-head` check what is the merge base between them and `master`. The list of merge bases may grow through time, but merge bases must not disappear, with the exception when there is a single (or no more then N, but better single) commit. Wondering whether travis cache may be used to get this behaviour.

---

Some discussion was going in #6431, but this does not look like a right place. Moving it here:

@ZyX-I
<blockquote>
The problem is that I did not intentionally touch anything which should cause E171, so I really do not know. And with git bisect you may understand why mercurial uses immutable history: I did not really make most of those commits, this is rebase. And while commit state was consistent when *I* was making a commit, after rebase it is no longer guaranteed. You have a bit better situation then it could possibly be because first few rebases I was checking whether tests are fine after solving conflicts.

Basically current development model keeps clean history, but is by no means good for any more or less big PR. E.g. even #6236 has inconsistent state in the middle of the PR: “Fix comments and cleanup tests” 3ea644c7c9e5c94d327fa6cf0b78470d2d0d2698 commit uses `rettv_dict_alloc`, but there is no such function and you can no longer bisect. And that is not a big PR.
</blockquote>

@justinmk 

<blockquote>
> Basically current development model keeps clean history, but is by no means good for any more or less big PR.

Would it help if we changed the recommendation to:

1. Cleanup ("squash fixup/noise commits") the feature branch on its _current_ origin (i.e. "common ancestor with upstream HEAD").
2. Then _merge_ to upstream HEAD, don't rebase.
</blockquote>

@ZyX-I

<blockquote>
I would suggest using `git ri` for interactive rebase to clean up `fixup!`s and merging upstream:

```
	ri = !zsh -c 'git rebase --interactive $(git merge-base --is-ancestor ${1:-master} ${2:-HEAD} && git rev-parse ${1:-master} || git merge-base ${1:-master} ${2:-HEAD}) ${@[3,-1]}' -
```

(usage: `git ri` if PR to master, `git ri X` if PR is supposed to be merged into X; never actually used $2).

Note: if I am not mistaking, `git ri` after merging will no longer allow altering post-merge commits (unless you use second argument). Though think this is a feature.

Alias is using zsh, but there are no zsh-ism apart from `${@[3, -1]}` and the fact that I did not bother with quoting (everything, but `$()` will appear in one piece in zsh), easy to fix.

If I am not mistaking, this is exactly what you mean by “rebase and squash the feature branch on its current origin”.

---

Additionally, it is fine to rebase the *single* commit: there will be no inconsistent state then.

---

No, this is the default behaviour. Used to determine what to rebase, not the destination.

---

After translating this to POSIX shell it became even more lengthy:

   	ri = "!sh -c 't=\"${1:-master}\"; s=\"${2:-HEAD}\" ; [ $# -gt 0 ] && shift ; [ $# -gt 0 ] && shift ; git rebase --interactive \"$(git merge-base --is-ancestor \"$t\" \"$s\" && git rev-parse \"$t\" || git merge-base \"$t\" \"$s\")\" \"$@\"' -"

(tested with `dash` in place of `sh`: my `sh` is `bash` symlink). It looks like on `#git` people either do not understand the request or have no idea how to implement it without such big script.
</blockquote>

**Comment by ZyX-I** [comment]
@justinmk Note: messages were removed from there.

**Comment by ZyX-I** [comment]
Updated comment with pros and cons.

**Comment by justinmk** [comment]
I'm going to try using your script.

As far as project policy, I would really hesitate to make any of this prescriptive. _Disallowing_ rebases by CI is drastic, and it also puts us on the hook for explaining the workflow and helping contributors understand it (or losing contributors...).

Until now we've actively discouraged merges. We can take the first step by changing the recommendation to (1) use the script, and (2) allowing merge-commits in large PRs.

For smallish PRs rebasing should continue to be fine.

**Comment by ZyX-I** [comment]
No script exists yet, but creating “disallow” part is going to be super easy (assuming travis cache works like I need): it consists of feeding `git merge-base` hashes from `git log master..feature` (need to pull in master if I am not mistaking) and comparing the output with the previous run.

Creating “script which generates a script which restores state unless user actively edited commit messages” part is not very hard as well: record last HEAD hash to cache, in a generated script restore branch pointer to that hash and do cherry-picking of new commits, and rerunning merge for new merge commits. “New” are determined based on commit messages (checked by commit message similarity, 90% or greater: in case typo amends). Easier to write in Python because computing similarity on bash… Of course, throws away conflict resolutions but `git rerere` should help here.

“Smallish” PRs are supposed to be single commit or squashing in this case: in any case alternative is a bunch of small commits and most of time history is perfectly fine with only one.

**Comment by ZyX-I** [comment]
Actually the script should be run as a part of lint job: git history is not going to change, there already are valid cases when this job failure is being ignored.

**Comment by jamessan** [comment]
I'm fine with not encouraging people to rebase, but I don't think we should state it's forbidden.  It should be up to the contributor and their development process.

The eval refactor changes didn't need to be rebased and if you felt that was required, then that's our fault for that perception.  The important parts (IMO) are keeping clean history (in the bisectable sense) and getting feedback to people in a timely manner so we reduce the amount of semantic conflicts that have to be dealt with.

Either way, there's no need to merge as soon as Git(Hub) shows there are conflicts.  Merges are only useful if they provide necessary content or are being done in preparation for delivering the PR.  People should focus on what they're trying to accomplish and merge lazily.

---
As far as rebases "causing" broken history, that's only the case if you don't check it.  You can trivially check each commit in a rebase by putting `exec make`/`exec make -C build` after each `pick`/`edit`/whatever line in an interactive rebase to verify that every commit still stands alone.

Rebasing really isn't as hard or as bad as people make it out to be.  The problem is when people focus on having a perfectly groomed history instead of just a functional history.

**Comment by ZyX-I** [comment]
Conflicts disabe CI (AFAIR AppVeyor was/is explicit about that, travis also has https://docs.travis-ci.com/user/pull-requests/#My-Pull-Request-isn%E2%80%99t-being-built (see last part)), so I *have* to solve them somehow right if I want to continue. It is the first time I heard merges are allowed.

`exec make …` after each commit are going to make rebases very time-consuming: need to run at least unit tests, better if all tests (not that I actually did that when I was checking). Though I really have missed this way to make solving this kind of problems automatically and times should go down should #6371 be fully implemented: now e.g. eval.c triggers almost complete rebuild even if eval functions are not used.

**Comment by jamessan** [comment]
> Conflicts disabe CI (AFAIR AppVeyor was/is explicit about that, travis also has https://docs.travis-ci.com/user/pull-requests/#My-Pull-Request-isn%E2%80%99t-being-built), so I have to solve them somehow right if I want to continue.

That's unfortunate, since it encourages either unnecessary merges or risky rebases.  I understand why the choice was made, but it makes longer term development more painful than it should be.

**Comment by ZyX-I** [comment]
So what we are going to do is

1. Document that history editing should preferrably use `ri` alias alongside with the alias.
2. Document that when rebasing it is suggested to add `exec make -C build && make -C build unittest` per one `pick`/`edit`/`reword` line; per one line but *after* all squashes and fixups, with the exception of single commits.
3. Document that big PRs are allowed to merge and should use merges because `exec`s are going to cause extensive problems in this case: there would be many and this is rather pointless.

And why I think merges not allowed: https://github.com/neovim/neovim/blob/master/CONTRIBUTING.md#pull-requests-prs talks about rebases, does not talk about merges. So I guess it should now be

---

- To avoid duplicate work, you may want to create a `[WIP]` pull request so that
  others know what you are working on.
- Avoid cosmetic changes to unrelated files in the same commit: extra noise
  makes reviews more difficult.
- Use a [feature branch][git-feature-branch] instead of the master branch.
- Edit history with `ri` alias: add

        [alias]
        	ri = "!sh -c 't=\"${1:-master}\"; s=\"${2:-HEAD}\" ; [ $# -gt 0 ] && shift ; [ $# -gt 0 ] && shift ; git rebase --interactive \"$(git merge-base --is-ancestor \"$t\" \"$s\" && git rev-parse \"$t\" || git merge-base \"$t\" \"$s\")\" \"$@\"' -"

    to your `~/.gitconfig`.

    This avoids unnecessary rebases yet still allows you to combine related commits, separate monolithic commits, etc.
- Rebase relatively small PRs, using `exec make -C build && make -C build unittest` after each pick/edit/reword, after all following squash/fixup.
- Merge in `master` of bigger PRs. Do not do excessive merging: merge in case of merge conflicts or when master introduces changes which break your PR.

    Do not edit commits which come before the merge commit.

**Comment by aktau** [comment]
Honestly, I agree most with what @jamessan said:

> As far as rebases "causing" broken history, that's only the case if you don't check it. You can trivially check each commit in a rebase by putting exec make/exec make -C build after each pick/edit/whatever line in an interactive rebase to verify that every commit still stands alone.
>
> Rebasing really isn't as hard or as bad as people make it out to be. The problem is when people focus on having a perfectly groomed history instead of just a functional history.

How many PRs do we have that contains 10s of commits? From a cursory glance, not too many. I think the onus should be on the person preparing the PR, and then we can have decent history too. (Maybe I'm underestimating that onus).

@ZyX-I said:

> exec make … after each commit are going to make rebases very time-consuming: need to run at least unit tests, better if all tests (not that I actually did that when I was checking). Though I really have missed this way to make solving this kind of problems automatically and times should go down should #6371 be fully implemented: now e.g. eval.c triggers almost complete rebuild even if eval functions are not used.

As you say, there are several degrees of checking. Actually doing `exec make ...` after each commit does not take a lot of time (that's how I did a lot of automated `git bisect skip`'s). This will at least take care of the compilation failures and allow some sort of bisecting.

Ideally one would run unit tests as well. How long does it take on an average laptop/workstation these days? If those are not parallelized, perhaps we can make a script to test multiple commits at once (though that would incur disk IO, likely even worse)? Or just parallelize them inline.

Even more, how long does it take to run the integration tests?

My machine here is really powerful, so I'm not going to assume everyone has something like that.

**Comment by ZyX-I** [comment]
@aktau Unit tests are generally less then a minute, especially with options that disable tracing and move headers parsing to the main process: this makes them run a bit faster then compilation. Main compilation due to broken incremental compilation is nearly the same, should total 1-2 mins per commit.

Functional tests are a lot worse: 13 minutes, this is why I did not suggest them. And what is bad is `cmake` run: if some of CMake files changed you will see that CMake doubles compilation time because it is doing something after reporting “configuring done”, but before reporting “generating done”.

The problem is that “configuring” depends on current git commit, so `run_unit_test` (shell function that calls busted directly) is 25 seconds with NVIM_TEST_MAIN_CDEFS=1 NVIM_TEST_TRACE_LEVEL=0, `make -C build nvim-test` without commit changes since last compilation is ~30 seconds (depending on how many things changed). Incremental compilation when *only* commit changed and not C code is ~1 minute: purely on CMake. As `make functionaltest` is doing one compilation and library is not too much different from binary I guess I should remove it from advice. But in any case this totals 1 minute for CMake regenerating, half a minute for compilation and another half for unit tests: two minutes total. This is half an hour wasted just on rebasing something like #6236 (unless author squashes which would be good for that PR). And don’t ask to run functional tests.

So have in CONTRIBUTING.md?

---

- To avoid duplicate work, you may want to create a `[WIP]` pull request so that
  others know what you are working on.
- Avoid cosmetic changes to unrelated files in the same commit: extra noise
  makes reviews more difficult.
- Use a [feature branch][git-feature-branch] instead of the master branch.
- Edit history with `ri` alias: add

        [alias]
        	ri = "!sh -c 't=\"${1:-master}\"; s=\"${2:-HEAD}\" ; [ $# -gt 0 ] && shift ; [ $# -gt 0 ] && shift ; git rebase --interactive \"$(git merge-base --is-ancestor \"$t\" \"$s\" && git rev-parse \"$t\" || git merge-base \"$t\" \"$s\")\" \"$@\"' -"

    to your `~/.gitconfig`.

    This avoids unnecessary rebases yet still allows you to combine related commits, separate monolithic commits, etc.
- Rebase relatively small PRs, using `exec make -C build unittest` after each pick/edit/reword, after all following squash/fixup.
- Merge in `master` of bigger PRs. Do not do excessive merging: merge in case of merge conflicts or when master introduces changes which break your PR.

    Do not edit commits which come before the merge commit.

**Comment by aktau** [comment]
> But in any case this totals 1 minute for CMake regenerating, half a minute for compilation and another half for unit tests: two minutes total. This is half an hour wasted just on rebasing something like #6236 

Is the CMake regeneration always necessary? For #6431 a build with `make clean && make ... install` takes ~10 seconds here (powerful workstation). Arguably, I shouldn't even have to do the clean, but I distrust make.

> (unless author squashes which would be good for that PR). And don’t ask to run functional tests.

I don't think it's always good to squash changes (unless the changes are fixups). For one, this would mean that the "culprit" commit found by bisect is one single huge commit. That's only a little bit more helpful than the current "can't bisect" situation. At least when I have commit summaries, I can guess.

> Do not edit commits which come before the merge commit.

Actually, now that I think of it, it would be great to have a `git` alias for: `git rebase -i HEAD..master`. (I'm not sure that's a legal command). Basically such that one cannot rewrite history, only your own commits.

I'm not a git master enough to see if that's what your `git ri` alias does.

**Comment by aktau** [comment]
(That said, I agree 30 minutes is a lot of time. But for preparing a huge PR for a merge, it may not be the worst thing ever. Huge PRs shouldn't be the norm.)

**Comment by ZyX-I** [comment]
> Is the CMake regeneration always necessary? For #6431 a build with make clean && make ... install takes ~10 seconds here (powerful workstation). Arguably, I shouldn't even have to do the clean, but I distrust make.

CMake regeneration is not “necessary”. It *happens*.

> Actually, now that I think of it, it would be great to have a git alias for: git rebase -i HEAD..master. (I'm not sure that's a legal command). Basically such that one cannot rewrite history, only your own commits.

> I'm not a git master enough to see if that's what your git ri alias does.

Not sure whether I understood you right. My `git ri` is there to edit history without rebasing: e.g. if you have A(B(C(D)), B'(C'(D'))) where `A` is `master` at the state when you started the PR, B/C/D are new commits in master and B'/C'/D' are commits in your PR after using `git ri` to squash C' and D' you should end up with A(B(C(D)), B'(E')). With `git rebase --interactive` you would end up with A(B(C(D(B'(E'))))).

**Comment by ZyX-I** [comment]
> (That said, I agree 30 minutes is a lot of time. But for preparing a huge PR for a merge, it may not be the worst thing ever. Huge PRs shouldn't be the norm.)

This is loosing initial context for almost no benefit. Basically the only cons which actually touches me is “history is less clean” in the meaning “history contains merge commits”: unless you do criss-cross merging (I hope nobody is ever going to do that) it is even still possible and easy to determine which commits come from the PR (git log `masterhead..PRhead`, where masterhead and PRhead are merge commit parents if I am not mistaking). Not something I would care much about and pros touch me all.

**Comment by aktau** [comment]
> Not sure whether I understood you right. My git ri is there to edit history without rebasing: e.g. if you have A(B(C(D)), B'(C'(D'))) where A is master at the state when you started the PR, B/C/D are new commits in master and B'/C'/D' are commits in your PR after using git ri to squash C' and D' you should end up with A(B(C(D)), B'(E')).

`A(B(C(D)), B'(E'))` sounds good, that's what I want. Never want to edit A(B(C(D)) (or basically, anything on master), since master is shared state.

In short: I want an alias so I can safely re-order/squash/edit any commits that I haven't officially shared (by pushing to master). This basically means: I want to only manage commits that are on my feature branches. Even more simply: I want to only manage commits that are: `Commits(current-branch) / Commits(master)` where `/` is the set difference operator.

> With git rebase --interactive you would end up with A(B(C(D(B'(E'))))).

This seems like it's both a rebase of the PR and then rebasing master on top of the rebased PR. `A(B(C(D(B'(E')))))` could flow out of `A(B(C(D)), B'(C'(D')))`.

What I don't want to happen is rewriting before master/HEAD, e.g., suppose the PR was branched from master/B:

```
// Before
master: A(B(C(D))
PR:       A(B(C'(D'))

// After (wrong rebase):
master: A(B(C(D))
PR:       A(B'(E') // B has turned into B' because I edited its commit message, or something.
```

Now the common parent is `A`, so I'm not sure `git rebase PR-branch` when on `master` is going to handle that cleanly.

**Comment by justinmk** [comment]
>  unless you do criss-cross merging (I hope nobody is ever going to do that)

IMO that would be something worth enforcing in CI.

**Comment by ZyX-I** [comment]
@aktau In both examples master is D and keeps being that, feature is D', then E'. `rebase --interactive` will not edit anything on master when doing a rebase of feature, but it will move the branch point from A to D, `ri` won’t.

**Comment by jamessan** [comment]
@aktau `git ri` is addressing the use case of simply rewriting commits since the last merge into your PR (or if no merges have occurred, since `@{upstream}`/`master`).  The intention is to provide people a means to merge master when necessary (if rebase isn't a good choice at that point) but have a common command to safely perform history editing.

The typical use of rebase -- `git rebase -i origin/master` -- doesn't allow for merging into your PR, but merging can help avoid creating non-bisectable history, since rebasing complicated and/or long-lived PRs can get messy.

**Comment by aktau** [comment]
Alright, I see now. Thanks @jamessan, @ZyX-I and @justinmk. I wasn't even considering the possibility of doing merge from master into the feature branch (and then back into master when the PR gets mainlined). I'm sure it'll work, though it's going to look a tad bit ugly.

**Comment by justinmk** [comment]
Closed in https://github.com/neovim/neovim/pull/6500
